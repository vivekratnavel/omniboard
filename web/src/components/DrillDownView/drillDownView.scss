$EXPANDED_ROW_HEIGHT: 500;
$HEADER_HEIGHT: 52;
$SIDEBAR_WIDTH: 200;
$light-steel-blue: #b0c4de;

@value EXPANDED_ROW_HEIGHT #{$EXPANDED_ROW_HEIGHT};

.full-height {
  height: 100%;
}

.drilldown-view {
  .sidebar-wrapper {
    height: #{$EXPANDED_ROW_HEIGHT - $HEADER_HEIGHT}px;
    border: 1px solid $light-steel-blue;

    .sidebar-container {
      float: left;
      width: #{$SIDEBAR_WIDTH}px;
      height: 100%;
      background-color: #F0FFFF;
      border-right: 1px solid #e2e2e2;
    }

    .sidebar-content {
      margin-left: #{$SIDEBAR_WIDTH}px;
      height: #{$EXPANDED_ROW_HEIGHT - $HEADER_HEIGHT - 5}px;;
      background-color: white;

      .tab-content {
        background-color: white;
        height: #{$EXPANDED_ROW_HEIGHT - $HEADER_HEIGHT - 5}px;;
        padding: 17px 20px;
        overflow: scroll;

        .metrics-plot-left {
          float: left;
          width: 250px;
        }

        .metrics-plot-content {
          margin-left: 250px;
        }
      }
    }
  }

  .checkbox, .radio {
    label {
      line-height: 20px;
    }
  }

  .navbar {
    margin: 0 !important;
    height: #{$HEADER_HEIGHT}px;
    border-radius: unset !important;

    &.navbar-light {
      background-color: #e3f2fd;
      border: 1px solid $light-steel-blue;
    }

    .navbar-brand {
      &:hover {
        cursor: pointer;
      }
    }
  }

  .nav-pills > li > a {
    border-radius: unset !important;
  }
}
