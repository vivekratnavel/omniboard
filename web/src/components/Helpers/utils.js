
export const capitalize = value => {
  // Capitalize each word
  let split = value.split('_');
  split = split.map(key => {
    if (key.length > 0) {
      return key.charAt(0).toUpperCase() + key.slice(1);
    }

    return '';
  });
  return split.join(' ');
};

export const reorderArray = (array, insertAfter, value) => {
  const insertAfterIndex = array.indexOf(insertAfter);
  const valueIndex = array.indexOf(value);
  if (valueIndex >= 0) {
    array.splice(valueIndex, 1);
    array.splice(insertAfterIndex + 1, 0, value);
  }
};

export const arrayDiff = (arr1, arr2) => arr1.filter(i => !arr2.includes(i));

export const arrayDiffColumns = (arr1, arr2) => arr1.filter(item => !arr2.some(i => i.name === item.name && i.value === item.value));

export const headerText = header => {
  return capitalize(header);
};

export const parseServerError = error => {
  const defaultMessage = 'Error: ';
  let message = '';
  if (error.response) {
    // The request was made and the server responded with a status code
    // that falls out of the range of 2xx
    message = error.response.data.message;
  } else if (error.request) {
    // The request was made but no response was received
    // `error.request` is an instance of XMLHttpRequest in the browser and an instance of
    // http.ClientRequest in node.js
    message = 'No response was received from the server.';
  } else {
    // Something happened in setting up the request that triggered an Error
    message = error.message;
  }

  return defaultMessage + message;
};

export const getFileExtension = fileName => fileName.split('.').splice(-1)[0];

/**
 * Creates a new ArrayBuffer from concatenating two existing ones
 *
 * @param {ArrayBuffer | null} buffer1 The first buffer.
 * @param {ArrayBuffer | null} buffer2 The second buffer.
 * @return {ArrayBuffer | null} The new ArrayBuffer created out of the two.
 */
export const concatArrayBuffers = (buffer1, buffer2) => {
  if (!buffer1) {
    return buffer2;
  }

  if (!buffer2) {
    return buffer1;
  }

  const tmp = new Uint8Array(buffer1.byteLength + buffer2.byteLength);
  tmp.set(new Uint8Array(buffer1), 0);
  tmp.set(new Uint8Array(buffer2), buffer1.byteLength);
  return tmp.buffer;
};

export const resolveObjectPath = (object, path, defaultValue) =>
  path.split('.').reduce((o, p) => o && Object.prototype.hasOwnProperty.call(o, p) ? o[p] : defaultValue, object);

export const getAllPaths = (prefix, data, excludeObjectType = false) => {
  return Object.keys(data).reduce((paths, key) => {
    const newPath = prefix ? `${prefix}.${key}` : key;
    if (!excludeObjectType || typeof data[key] !== 'object') {
      paths = [...paths, newPath];
    }

    if (typeof data[key] === 'object' && !Array.isArray(data[key]) && data[key]) {
      const newPaths = Object.keys(data[key]).reduce((acc, item) => {
        if (typeof data[key][item] !== 'object' && data[key][item]) {
          acc.push(`${newPath}.${item}`);
        }

        return acc;
      }, []);
      const recursivePaths = getAllPaths(newPath, data[key], excludeObjectType);
      return [...paths, ...recursivePaths, ...newPaths];
    }

    return paths;
  }, []);
};

export const getOption = (value, options) => {
  return options.find(option => option.value === value);
};
