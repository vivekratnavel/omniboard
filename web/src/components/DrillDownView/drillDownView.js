import PropTypes from 'prop-types';
import React, {Component} from 'react';
import axios from 'axios';
import ReactJson from 'react-json-view';
import './drillDownView.scss';
import {Nav, NavItem, Glyphicon} from 'react-bootstrap';
import {on, off} from 'dom-helpers/events';
import LocalStorageMixin from 'react-localstorage';
import reactMixin from 'react-mixin';
import {toast} from 'react-toastify';
import backend, {setDbInfo} from '../Backend/backend';
import {ProgressWrapper} from '../Helpers/hoc';
import {parseServerError} from '../Helpers/utils';
import {DRILLDOWN_VIEW} from '../../appConstants/drillDownView.constants';
import {MetricsPlotView} from '../MetricsPlotView/metricsPlotView';
import {SourceFilesView} from '../SourceFilesView/sourceFilesView';
import {CapturedOutView} from '../CapturedOutView/capturedOutView';
import {STATUS} from '../../appConstants/status.constants';

class JsonView extends React.PureComponent {
  static propTypes = {
    data: PropTypes.object
  };

  render() {
    const {data} = this.props;
    return (
      <ReactJson src={data} iconStyle='circle' collapsed={1}
        collapseStringsAfterLength={40}
        enableClipboard={false} displayDataTypes={false}/>
    );
  }
}

class DrillDownView extends Component {
  static propTypes = {
    height: PropTypes.number,
    runId: PropTypes.number.isRequired,
    width: PropTypes.number,
    status: PropTypes.string.isRequired,
    dbInfo: PropTypes.shape({
      path: PropTypes.string,
      key: PropTypes.string
    }).isRequired,
    handleExpandViewClick: PropTypes.func,
    showHeader: PropTypes.bool
  };

  // Filter out state objects that need to be synchronized with local storage
  static defaultProps = {
    stateFilterKeys: ['selectedNavTab']
  };

  constructor(props) {
    super(props);
    this.state = {
      selectedNavTab: DRILLDOWN_VIEW.METRICS,
      isTableLoading: false,
      runsResponse: null,
      metricsResponse: null
    };
  }

  _handleSelectNavPill = selectedKey => {
    this.setState({
      selectedNavTab: selectedKey
    });
  };

  _stopWheel = event => {
    const {scrollTop, scrollHeight, clientHeight} = this.scrollableDiv;
    const maxScrollTop = scrollHeight - clientHeight;
    const scrollingUp = event.deltaY < 0;
    const scrollingDown = event.deltaY > 0;
    // Handle event only if scrolling up or down
    if (scrollingUp || scrollingDown) {
      // Stop event propagation to allow DrillDownView content to scroll
      if ((scrollTop > 0 && (scrollTop < maxScrollTop || scrollingUp)) || (scrollTop === 0 && maxScrollTop > 0 && scrollingDown)) {
        event.stopPropagation();
      }
    }

    if (this.state.selectedNavTab === DRILLDOWN_VIEW.CAPTURED_OUT) {
      // Stop event propagation for horizontal scroll to allow CAPTURED_OUT content to be scrolled
      if (event.deltaX !== 0) {
        event.stopPropagation();
      }
    }
  };

  componentDidMount() {
    const {runId, dbInfo} = this.props;
    setDbInfo(backend, dbInfo);
    if (runId) {
      this.setState({
        isTableLoading: true
      });
      axios.all([
        backend.get(`api/v1/Runs/${runId}`, {
          params: {
            select: 'captured_out,info,meta,host,experiment,artifacts,config,fail_trace,status'
          }
        }),
        backend.get('api/v1/Metrics', {
          params: {
            query: `{"run_id":"${runId}"}`
          }
        })
      ])
        .then(axios.spread((runsResponse, metricsResponse) => {
          const runsResponseData = runsResponse.data;
          const metricsResponseData = metricsResponse.data;
          this.setState({
            isTableLoading: false,
            runsResponse: runsResponseData,
            metricsResponse: metricsResponseData
          });
        }))
        .catch(error => {
          this.setState({isTableLoading: false});
          toast.error(parseServerError(error));
        });
    }

    on(this.scrollableDiv, 'wheel', this._stopWheel);
  }

  componentWillUnmount() {
    off(this.scrollableDiv, 'wheel', this._stopWheel);
  }

  render() {
    const {selectedNavTab, isTableLoading, runsResponse, metricsResponse} = this.state;
    const {width, height, runId, status, dbInfo, handleExpandViewClick, showHeader} = this.props;
    const experimentName = runsResponse !== null && 'experiment' in runsResponse ? runsResponse.experiment.name : '';
    // Adjust the width of scrollbar from total width
    const style = {
      height,
      width: width - 17
    };
    const localStorageKey = `${dbInfo.key}|MetricsPlotView|${runId}`;
    let experimentFiles = [];
    // Convert experiment sources to the same structure as artifacts
    // source is an array with the following structure
    // [ "filename", ObjectId ]
    if (runsResponse !== null && 'experiment' in runsResponse &&
      runsResponse.experiment.sources && runsResponse.experiment.sources.length > 0) {
      experimentFiles = runsResponse.experiment.sources.map(source => {
        return {
          name: source[0],
          file_id: source[1]
        };
      });
    }

    const showFailTrace = runsResponse && runsResponse.status === STATUS.FAILED;
    const failTraceMenuItem = showFailTrace ?
      <NavItem eventKey={DRILLDOWN_VIEW.FAIL_TRACE}>Fail Trace</NavItem> :
      null;
    const failTrace = runsResponse && runsResponse.fail_trace ?
      runsResponse.fail_trace.reduce((accumulator, current) => accumulator + current + '\n', '') :
      '';
    const failTraceDom = <pre>{failTrace}</pre>;

    const getTabContent = key => {
      let content = '';
      if (runsResponse) {
        switch (key) {
          case DRILLDOWN_VIEW.EXPERIMENT:
            content = <div id={DRILLDOWN_VIEW.EXPERIMENT}><JsonView data={runsResponse.experiment}/></div>;
            break;
          case DRILLDOWN_VIEW.HOST_INFO:
            content = <div id={DRILLDOWN_VIEW.HOST_INFO}><JsonView data={runsResponse.host}/></div>;
            break;
          case DRILLDOWN_VIEW.RUN_INFO:
            content = <div id={DRILLDOWN_VIEW.RUN_INFO}><JsonView data={runsResponse.info}/></div>;
            break;
          case DRILLDOWN_VIEW.CAPTURED_OUT:
            content = <div id={DRILLDOWN_VIEW.CAPTURED_OUT}><CapturedOutView initialOutput={runsResponse.captured_out} initialStatus={status} runId={runId}/></div>;
            break;
          case DRILLDOWN_VIEW.FAIL_TRACE:
            content = <div id={DRILLDOWN_VIEW.FAIL_TRACE}>{failTraceDom}</div>;
            break;
          case DRILLDOWN_VIEW.META_INFO:
            content = <div id={DRILLDOWN_VIEW.META_INFO}><JsonView data={runsResponse.meta}/></div>;
            break;
          case DRILLDOWN_VIEW.METRICS:
            content = <div id={DRILLDOWN_VIEW.METRICS}><MetricsPlotView metricsResponse={metricsResponse} runId={runId} dbInfo={dbInfo} localStorageKey={localStorageKey}/></div>;
            break;
          case DRILLDOWN_VIEW.ARTIFACTS:
            content = (
              <div id={DRILLDOWN_VIEW.ARTIFACTS}>
                <SourceFilesView key={'artifacts-' + runsResponse._id} type='artifacts' dbInfo={dbInfo} runId={runsResponse._id} files={runsResponse.artifacts}/>
              </div>
            );
            break;
          case DRILLDOWN_VIEW.SOURCE_FILES:
            content = (
              <div id={DRILLDOWN_VIEW.SOURCE_FILES}>
                <SourceFilesView key={'files-' + runsResponse._id} type='source_files' dbInfo={dbInfo} runId={runsResponse._id} files={experimentFiles}/>
              </div>
            );
            break;
          case DRILLDOWN_VIEW.CONFIG:
            content = <div id={DRILLDOWN_VIEW.CONFIG}><JsonView data={runsResponse.config}/></div>;
            break;
          default:
        }
      }

      return <ProgressWrapper id='ddv-progress-wrapper' loading={isTableLoading}>{content}</ProgressWrapper>;
    };

    const getHeaderContent = showHeader ? (
      <nav className='navbar navbar-light'>
        <span className='navbar-brand'>
          <a onClick={() => handleExpandViewClick(runId, status)}>
            Details for: <strong>{experimentName}</strong> (Id: {runId}) &nbsp;
            <Glyphicon glyph='resize-full'/>
          </a>
        </span>
      </nav>
    ) : null;

    return (
      <div style={style} className='drilldown-view'>
        {getHeaderContent}
        <div className='sidebar-wrapper'>
          <div className='sidebar-container full-height'>
            <div id='sidebar'>
              <Nav stacked bsStyle='pills' activeKey={selectedNavTab} onSelect={this._handleSelectNavPill}>
                <NavItem eventKey={DRILLDOWN_VIEW.METRICS}>
                  Metrics Plot
                </NavItem>
                <NavItem eventKey={DRILLDOWN_VIEW.CAPTURED_OUT}>
                  Captured Out
                </NavItem>
                {failTraceMenuItem}
                <NavItem eventKey={DRILLDOWN_VIEW.EXPERIMENT}>
                  Experiment Details
                </NavItem>
                <NavItem eventKey={DRILLDOWN_VIEW.HOST_INFO}>
                  Host Info
                </NavItem>
                <NavItem eventKey={DRILLDOWN_VIEW.RUN_INFO}>
                  Run Info
                </NavItem>
                <NavItem eventKey={DRILLDOWN_VIEW.META_INFO}>
                  Meta Info
                </NavItem>
                <NavItem eventKey={DRILLDOWN_VIEW.ARTIFACTS}>
                  Artifacts
                </NavItem>
                <NavItem eventKey={DRILLDOWN_VIEW.SOURCE_FILES}>
                  Source Files
                </NavItem>
                <NavItem eventKey={DRILLDOWN_VIEW.CONFIG}>
                  Config
                </NavItem>
              </Nav>
            </div>
          </div>
          <div className='sidebar-content full-height'>
            <div ref={div => (this.scrollableDiv = div || this.scrollableDiv)} className='tab-content'>
              {getTabContent(selectedNavTab)}
            </div>
          </div>
        </div>
      </div>
    );
  }
}

reactMixin(DrillDownView.prototype, LocalStorageMixin);
export {DrillDownView};
