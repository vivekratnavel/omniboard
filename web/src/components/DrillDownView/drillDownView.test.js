import React from 'react';
import mockAxios from 'jest-mock-axios';
import {toast} from 'react-toastify';
import * as events from 'dom-helpers/events';
import {DRILLDOWN_VIEW} from '../../appConstants/drillDownView.constants';
import {parseServerError} from '../Helpers/utils';
import {STATUS} from '../../appConstants/status.constants';
import {DrillDownView} from './drillDownView';

describe('DrillDownView', () => {
  let wrapper = null;
  const off = jest.spyOn(events, 'off');
  console.error = jest.fn();
  toast.error = jest.fn();

  const dbInfo = {key: 'default', name: 'test_db'};

  const runsResponseData = {
    _id: 200,
    meta: {},
    info: {},
    host: {},
    experiment: {
      sources: [
        ['hello_world.py', 'SGVsbG8gV29ybGQh']
      ]
    },
    artifacts: [],
    captured_out: 'captured out',
    status: STATUS.COMPLETED
  };

  const runsResponseDataForFailed = {
    _id: 200,
    meta: {},
    info: {},
    host: {},
    experiment: {
      sources: [
        ['hello_world.py', 'SGVsbG8gV29ybGQh']
      ]
    },
    artifacts: [],
    captured_out: 'captured out',
    status: STATUS.FAILED,
    fail_trace: ['Traceback (most recent call last):\n', '  File "/Users/anaconda3/lib/python3.5/site-packages/sacred/config/captured_function.py", line 46, in captured_function\n    result = wrapped(*args, **kwargs)\n', '  File "hello.py", line 55, in my_main\n    svg2png(bytestring=dwg.tostring(), write_to=\'output.png\')\n']
  };

  const metricsResponseData = [
    {_id: '5a2b5a8c9c7a505a652f6127', name: 'pretrain.train.loss', run_id: 222, steps: [0, 1, 2, 3, 4],
      values: [0.7159541824544438, 0.3840367944955761, 0.3469185283233073, 0.30483262065173106, 0.28915774130337507],
      timestamps: ['2017-12-09T03:37:44.425Z', '2017-12-09T03:41:54.414Z', '2017-12-09T03:46:01.766Z', '2017-12-09T03:50:07.365Z', '2017-12-09T03:54:12.560Z']},
    {_id: '5a2b5aa09c7a505a652f6146', name: 'pretrain.val.loss', run_id: 223, steps: [0, 1, 2, 3, 4],
      values: [0.32177006650114165, 0.23237958704995795, 0.23340759051386187, 0.21925230575196739, 0.20541178824900605],
      timestamps: ['2017-12-09T03:38:01.945Z', '2017-12-09T03:42:11.673Z', '2017-12-09T03:46:18.843Z', '2017-12-09T03:50:24.377Z', '2017-12-09T03:54:29.752Z']}
  ];

  beforeEach(() => {
    wrapper = mount(
      <DrillDownView showHeader height={500} runId={200} width={600} status={STATUS.COMPLETED} dbInfo={dbInfo}/>
    );
  });

  afterEach(() => {
    // Cleaning up the mess left behind the previous test
    mockAxios.reset();
    jest.clearAllMocks();
  });

  it('should render correctly', async () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.state().isTableLoading).toBeTruthy();
    mockAxios.mockResponse({status: 200, data: runsResponseData});
    mockAxios.mockResponse({status: 200, data: metricsResponseData});
    wrapper.setState({selectedNavTab: DRILLDOWN_VIEW.EXPERIMENT});
    await tick();

    expect(wrapper.state().isTableLoading).toBeFalsy();
  });

  it('should switch tabs correctly', async () => {
    mockAxios.mockResponse({status: 200, data: runsResponseData});
    mockAxios.mockResponse({status: 200, data: metricsResponseData});
    wrapper.find('a').last().simulate('click');

    expect(wrapper.find('li').last().hasClass('active')).toBeTruthy();
    expect(wrapper.find('li.active')).toHaveLength(1);
    await tick();
    const switchTabTest = tabKey => {
      wrapper.setState({selectedNavTab: tabKey});

      expect(wrapper.find('div.tab-content ProgressWrapper#ddv-progress-wrapper').instance().props.children.props.id).toEqual(tabKey);
    };

    [
      DRILLDOWN_VIEW.EXPERIMENT,
      DRILLDOWN_VIEW.CAPTURED_OUT,
      DRILLDOWN_VIEW.HOST_INFO,
      DRILLDOWN_VIEW.META_INFO,
      DRILLDOWN_VIEW.METRICS,
      DRILLDOWN_VIEW.RUN_INFO,
      DRILLDOWN_VIEW.ARTIFACTS,
      DRILLDOWN_VIEW.SOURCE_FILES
    ].forEach(key => switchTabTest(key));
  });

  it('should switch to fail trace tab for failed run', async () => {
    mockAxios.mockResponse({status: 200, data: runsResponseDataForFailed});
    mockAxios.mockResponse({status: 200, data: metricsResponseData});

    await tick();
    wrapper.setState({selectedNavTab: DRILLDOWN_VIEW.FAIL_TRACE});

    expect(wrapper.find('div.tab-content ProgressWrapper#ddv-progress-wrapper').instance().props.children.props.id).toEqual(DRILLDOWN_VIEW.FAIL_TRACE);
  });

  it('should handle errors', async () => {
    const err = {status: 500, message: 'internal server error'};
    mockAxios.mockResponse({status: 200, data: runsResponseData});
    mockAxios.mockError(err);
    await tick();

    expect(toast.error).toHaveBeenCalledWith(parseServerError(err));
  });

  describe('should stop mouse wheel propagation', () => {
    beforeEach(() => {
      wrapper.setState({selectedNavTab: DRILLDOWN_VIEW.EXPERIMENT});
    });

    it('for horizontal scroll', async () => {
      mockAxios.mockResponse({status: 200, data: runsResponseData});
      mockAxios.mockResponse({status: 200, data: metricsResponseData});
      const event = {deltaX: 5, stopPropagation: jest.fn()};
      wrapper.instance()._stopWheel(event);
      await tick();

      expect(event.stopPropagation).not.toHaveBeenCalled();
      wrapper.setState({selectedNavTab: DRILLDOWN_VIEW.CAPTURED_OUT});
      wrapper.instance()._stopWheel(event);

      expect(event.stopPropagation).toHaveBeenCalledWith();
    });

    it('for vertical scroll', async () => {
      mockAxios.mockResponse({status: 200, data: runsResponseData});
      mockAxios.mockResponse({status: 200, data: metricsResponseData});
      const event = {deltaY: -5, stopPropagation: jest.fn()};

      expect(event.stopPropagation).toHaveBeenCalledTimes(0);
      wrapper.instance()._stopWheel(event);
      await tick();

      expect(event.stopPropagation).not.toHaveBeenCalled();
      wrapper.instance().scrollableDiv = {scrollTop: 5, scrollHeight: 30, clientHeight: 40};
      wrapper.instance()._stopWheel(event);

      expect(event.stopPropagation).toHaveBeenCalledTimes(1);
      wrapper.instance().scrollableDiv = {scrollTop: 0, scrollHeight: 100, clientHeight: 40};
      event.deltaY = 5;
      wrapper.instance()._stopWheel(event);

      expect(event.stopPropagation).toHaveBeenCalledTimes(2);
    });
  });

  it('unregisters event listener on unmount', () => {
    expect(off).not.toHaveBeenCalled();
    wrapper.unmount();

    expect(off).toHaveBeenCalledTimes(1);
  });

  it('should not call API without runId', () => {
    wrapper.unmount();
    mockAxios.reset();
    wrapper = mount(
      <DrillDownView height={500} width={600} runId={0} status={STATUS.COMPLETED} dbInfo={dbInfo}/>
    );

    expect(mockAxios.get).not.toHaveBeenCalled();
    wrapper.unmount();
    wrapper = mount(
      <DrillDownView height={500} runId={1} width={600} status={STATUS.COMPLETED} dbInfo={dbInfo}/>
    );

    expect(mockAxios.get).toHaveBeenCalledTimes(2);
  });
});
