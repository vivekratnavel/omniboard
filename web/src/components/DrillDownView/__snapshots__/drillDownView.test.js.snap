// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrillDownView should render correctly 1`] = `
<DrillDownView
  dbInfo={
    Object {
      "key": "default",
      "name": "test_db",
    }
  }
  height={500}
  runId={200}
  showHeader={true}
  stateFilterKeys={
    Array [
      "selectedNavTab",
    ]
  }
  status="COMPLETED"
  width={600}
>
  <div
    className="drilldown-view"
    style={
      Object {
        "height": 500,
        "width": 583,
      }
    }
  >
    <nav
      className="navbar navbar-light"
    >
      <span
        className="navbar-brand"
      >
        <a
          onClick={[Function]}
        >
          Details for: 
          <strong />
           (Id: 
          200
          ) Â 
          <Glyphicon
            bsClass="glyphicon"
            glyph="resize-full"
          >
            <span
              className="glyphicon glyphicon-resize-full"
            />
          </Glyphicon>
        </a>
      </span>
    </nav>
    <div
      className="sidebar-wrapper"
    >
      <div
        className="sidebar-container full-height"
      >
        <div
          id="sidebar"
        >
          <Nav
            activeKey="metrics"
            bsClass="nav"
            bsStyle="pills"
            justified={false}
            onSelect={[Function]}
            pullLeft={false}
            pullRight={false}
            stacked={true}
          >
            <ul
              className="nav nav-pills nav-stacked"
              role={null}
            >
              <NavItem
                active={true}
                activeKey="metrics"
                disabled={false}
                eventKey="metrics"
                key=".0"
                onSelect={[Function]}
              >
                <li
                  className="active"
                  role="presentation"
                >
                  <SafeAnchor
                    componentClass="a"
                    disabled={false}
                    onClick={[Function]}
                  >
                    <a
                      href="#"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="button"
                    >
                      Metrics Plot
                    </a>
                  </SafeAnchor>
                </li>
              </NavItem>
              <NavItem
                active={false}
                activeKey="metrics"
                disabled={false}
                eventKey="captured_out"
                key=".1"
                onSelect={[Function]}
              >
                <li
                  className=""
                  role="presentation"
                >
                  <SafeAnchor
                    componentClass="a"
                    disabled={false}
                    onClick={[Function]}
                  >
                    <a
                      href="#"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="button"
                    >
                      Captured Out
                    </a>
                  </SafeAnchor>
                </li>
              </NavItem>
              <NavItem
                active={false}
                activeKey="metrics"
                disabled={false}
                eventKey="experiment"
                key=".3"
                onSelect={[Function]}
              >
                <li
                  className=""
                  role="presentation"
                >
                  <SafeAnchor
                    componentClass="a"
                    disabled={false}
                    onClick={[Function]}
                  >
                    <a
                      href="#"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="button"
                    >
                      Experiment Details
                    </a>
                  </SafeAnchor>
                </li>
              </NavItem>
              <NavItem
                active={false}
                activeKey="metrics"
                disabled={false}
                eventKey="host_info"
                key=".4"
                onSelect={[Function]}
              >
                <li
                  className=""
                  role="presentation"
                >
                  <SafeAnchor
                    componentClass="a"
                    disabled={false}
                    onClick={[Function]}
                  >
                    <a
                      href="#"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="button"
                    >
                      Host Info
                    </a>
                  </SafeAnchor>
                </li>
              </NavItem>
              <NavItem
                active={false}
                activeKey="metrics"
                disabled={false}
                eventKey="run_info"
                key=".5"
                onSelect={[Function]}
              >
                <li
                  className=""
                  role="presentation"
                >
                  <SafeAnchor
                    componentClass="a"
                    disabled={false}
                    onClick={[Function]}
                  >
                    <a
                      href="#"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="button"
                    >
                      Run Info
                    </a>
                  </SafeAnchor>
                </li>
              </NavItem>
              <NavItem
                active={false}
                activeKey="metrics"
                disabled={false}
                eventKey="meta_info"
                key=".6"
                onSelect={[Function]}
              >
                <li
                  className=""
                  role="presentation"
                >
                  <SafeAnchor
                    componentClass="a"
                    disabled={false}
                    onClick={[Function]}
                  >
                    <a
                      href="#"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="button"
                    >
                      Meta Info
                    </a>
                  </SafeAnchor>
                </li>
              </NavItem>
              <NavItem
                active={false}
                activeKey="metrics"
                disabled={false}
                eventKey="artifacts"
                key=".7"
                onSelect={[Function]}
              >
                <li
                  className=""
                  role="presentation"
                >
                  <SafeAnchor
                    componentClass="a"
                    disabled={false}
                    onClick={[Function]}
                  >
                    <a
                      href="#"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="button"
                    >
                      Artifacts
                    </a>
                  </SafeAnchor>
                </li>
              </NavItem>
              <NavItem
                active={false}
                activeKey="metrics"
                disabled={false}
                eventKey="source_files"
                key=".8"
                onSelect={[Function]}
              >
                <li
                  className=""
                  role="presentation"
                >
                  <SafeAnchor
                    componentClass="a"
                    disabled={false}
                    onClick={[Function]}
                  >
                    <a
                      href="#"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="button"
                    >
                      Source Files
                    </a>
                  </SafeAnchor>
                </li>
              </NavItem>
              <NavItem
                active={false}
                activeKey="metrics"
                disabled={false}
                eventKey="config"
                key=".9"
                onSelect={[Function]}
              >
                <li
                  className=""
                  role="presentation"
                >
                  <SafeAnchor
                    componentClass="a"
                    disabled={false}
                    onClick={[Function]}
                  >
                    <a
                      href="#"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="button"
                    >
                      Config
                    </a>
                  </SafeAnchor>
                </li>
              </NavItem>
            </ul>
          </Nav>
        </div>
      </div>
      <div
        className="sidebar-content full-height"
      >
        <div
          className="tab-content"
        >
          <ProgressWrapper
            id="ddv-progress-wrapper"
            loading={true}
          >
            <div
              className="table-display"
            >
              <div
                className="table-cell-display gray"
              >
                <i
                  className="glyphicon glyphicon-refresh glyphicon-refresh-animate"
                />
                 Loading...
              </div>
            </div>
          </ProgressWrapper>
        </div>
      </div>
    </div>
  </div>
</DrillDownView>
`;
